dotnet/src/InternalUtilities/src/Diagnostics/ModelDiagnostics.cs

```diff
@@ -0,0 +1,246 @@
+        => StartCompletionActivity(endpoint, modelName, modelProvider, chatHistory, executionSettings, chatHistory => ToOpenAIFormat(chatHistory.AsEnumerable()));
+
+    /// <summary>
+    /// Set the text completion response for a given activity.
+    /// The activity will be enriched with the response attributes specified by the semantic conventions.
+    /// </summary>
+    public static void SetCompletionResponse(this Activity activity, IEnumerable<TextContent> completions, int? promptTokens, int? completionTokens)
+        => SetCompletionResponse(activity, completions, promptTokens, completionTokens, completions => $"[{string.Join(", ", completions)}]");
+
+    /// <summary>
+    /// Set the chat completion response for a given activity.
+    /// The activity will be enriched with the response attributes specified by the semantic conventions.
+    /// Token counts will be set to -1 if not provided.
+    /// </summary>
+    public static void SetCompletionResponse(this Activity activity, IEnumerable<TextContent> completions)
+            => SetCompletionResponse(activity, completions, null, null, completions => $"[{string.Join(", ", completions)}]");
+
+    /// <summary>
+    /// Set the chat completion response for a given activity.
+    /// The activity will be enriched with the response attributes specified by the semantic conventions.
+    /// </summary>
+    public static void SetCompletionResponse(this Activity activity, IEnumerable<ChatMessageContent> completions, int? promptTokens, int? completionTokens)
+        => SetCompletionResponse(activity, completions, promptTokens, completionTokens, ToOpenAIFormat);
+
+    /// <summary>
+    /// Set the chat completion response for a given activity.
+    /// The activity will be enriched with the response attributes specified by the semantic conventions.
+    /// Token counts will be set to -1 if not provided.
+    /// </summary>
+    public static void SetCompletionResponse(this Activity activity, IEnumerable<ChatMessageContent> completions)
+        => SetCompletionResponse(activity, completions, null, null, ToOpenAIFormat);
+
+    # region Private
+    /// <summary>
+    /// Check if model diagnostics is enabled
+    /// Model diagnostics is enabled if either EnableModelDiagnostics or EnableSensitiveEvents is set to true.
+    /// </summary>
+    private static bool IsModelDiagnosticsEnabled()
+    {
+        return s_enableModelDiagnostics || s_enableSensitiveEvents;
+    }
+
+    private static void AddOptionalTags(Activity? activity, PromptExecutionSettings? executionSettings)
+    {
+        if (activity is null || executionSettings?.ExtensionData is null)
+        {
+            return;
+        }
+
+        void TryAddTag(string key, string tag)
+        {
+            if (executionSettings.ExtensionData.TryGetValue(key, out var value))
+            {
+                activity.AddTag(tag, value);
```

Comment:
```suggestion
                activity.SetTag(tag, value);
```

Adding a tag supports duplication, setting a tag overwrites the existing (if any) value.
We made a mistake in the past and accidentally added the same attribute twice which broke some parts of the world. So `SetTag` is always safer :)